src/fairseq2/datasets/asr.py

```diff
@@ -92,160 +75,172 @@ def splits(self) -> set[str]:
+        )
 
-            builder.filter(skip)
+    @staticmethod
+    def add_tokenization_pipeline(
+        builder: DataPipelineBuilder,
+        tokenizer: TextTokenizer,
+        options: SpeechReadOptions | None = None,
+    ) -> DataPipelineBuilder:
+        if options is None:
+            options = SpeechReadOptions()
 
-            # Bucket `batch_size` examples.
-            builder.bucket(batching.batch_size, drop_remainder=options.drop_remainder)
-        else:
-            raise NotSupportedError(f"`{batching}` is not supported.")
+        # Tokenize target text.
+        text_encoder = tokenizer.create_encoder()
 
-        # Shuffle buckets.
-        if options.batch_shuffle_window != 1:
-            builder.shuffle(options.batch_shuffle_window, seed)
+        # to avoid to tokenize empty text, we filter out them out first
+        builder = builder.filter(lambda x: bool(len(x["text"]) > 0))
 
-        seed += 1
+        builder.map(text_encoder, selector="text")
 
-        # Memory map audio files.
-        cached_fd_count = options.extras.get("cached_fd_count", 1)
-        if not isinstance(cached_fd_count, int):
-            raise TypeError(
-                f"`options.extras['cached_fd_count']` must be of type `int`, but is of type `{type(cached_fd_count)}` instead."
-            )
+        unk_idx = tokenizer.vocab_info.unk_idx
 
-        file_mapper = FileMapper(audio_dir, cached_fd_count=cached_fd_count)
+        def empty_text(example: Dict[str, Any]) -> bool:
+            return bool((example["text"] != unk_idx).sum().item() > 0)
 
-        builder.map(file_mapper, selector="[*].audio")
+        if unk_idx is not None:
+            builder = builder.filter(empty_text)
 
-        # Decode audio.
-        audio_decoder = AudioDecoder(
-            dtype=torch.float32 if options.normalize_audio else options.dtype
+        filter_long_text_threshold = options.extras.get(
+            "filter_long_text_threshold", None
         )
+        if filter_long_text_threshold is not None:
+
+            def filter_long_text(example: Dict[str, Any]) -> bool:
+                return bool(example["text"].numel() <= filter_long_text_threshold)
 
-        builder.map(audio_decoder, selector="[*].audio.data")
+            builder = builder.filter(filter_long_text)
 
-        # TODO(balioglu): Check/adjust sample size
+        if options is not None:
+            remove_unknown = options.extras.get("remove_unknown", False)
+        else:
+            remove_unknown = False
 
-        # Normalize audio if requested.
-        def normalize(waveform: Tensor) -> Tensor:
-            with torch.no_grad():
-                waveform = layer_norm(waveform, waveform.shape)
+        assert isinstance(remove_unknown, bool)
+
+        if remove_unknown:
+            builder = builder.map(
+                lambda tensor: tensor[tensor != unk_idx], selector="text"
+            )
 
-            return waveform.to(options.dtype)
+        return builder
```

Comment:
This entire file is already upgraded to v0.5 (see [here](https://github.com/facebookresearch/fairseq2/blob/wav2vec2_trainer/recipes/wav2vec2/asr/data/dataset.py), but the selected code is missing. I also took a bit of liberty of grouping the code slightly more verbosely to enable new users to understand the fundamentals. Very happy to group/structure it differently. 

I will update this to fit the new mold of using specific configs for these filtering options. 

The biggest difference in the way of thinking of this file is that this tried to be as generic as possible with the ManifestInterface, but this is not necessary anymore as datasets are not required to be generic anymore - we can simply create them for specific recipes now which removes additional misdirection through interfaces.